File: src/main/java/cc/squirtle/Entity/CmdResult.java
package cc.squirtle.Entity;

import java.util.Collection;
import java.util.Random;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import net.md_5.bungee.api.ChatColor;


public class CmdResult<T> {
    private int type;
    private String title;
    private T data;

    //get server all setting info
    //public static App app;

    protected CmdResult(){

    }
    protected CmdResult(int type, String title, T data) {
        this.type = type;
        this.title = title;
        this.data = data;
    }

    /**
     * 成功返回结果
     *
     * @param data 获取的数据
     */
    public static <T> CmdResult<T> SUCCESS(T data) {
        return new CmdResult<T>(ResType.SUCCESS.getType(), ResType.SUCCESS.getTitle(), data);
    }

    public static <T> CmdResult<T> FAILED(T data) {
        return new CmdResult<T>(ResType.FAILED.getType(), ResType.FAILED.getTitle(), data);
    }
    public static <T> CmdResult<T> NOTICE(T data) {
        return new CmdResult<T>(ResType.NOTICE.getType(), ResType.NOTICE.getTitle(), data);
    }

    public static <T> CmdResult<T> INFO(T data) {
        return new CmdResult<T>(ResType.INFO.getType(), ResType.INFO.getTitle(), data);
    }
    public static <T> CmdResult<T> NAME(T data) {

        return new CmdResult<T>(ResType.NAME.getType(), ResType.NAME.getTitle(), data);
    }

    public static <T> CmdResult<T> UNAVAIL(T data) {
        return new CmdResult<T>(ResType.UNAVAIL.getType(), ResType.UNAVAIL.getTitle(), data);
    }
    public static <T> CmdResult<T> RAINBOW(T data) {
        return new CmdResult<T>(ResType.RAINBOW.getType(), ResType.RAINBOW.getTitle(), data);
    }

    /**
     * print message to console
     * @param cmdResult
     */
    public static void ConsolePrint(CmdResult<String> cmdResult) {
        // result name 
        String result = cmdResult.toString();
        if(cmdResult.type ==  203) {
            //GetPluginApp().getDescription().getName()
            result = result.replaceAll("%name%", PluginEntity.getPluginDescrption().getName());
        }else if(cmdResult.type == 102){
            result = RanbowString(result);
        }
 
        Bukkit.getConsoleSender().sendMessage(ChatColor.translateAlternateColorCodes('&', result));
    }

    /**
     * send Message to single-player
     * @param player
     * @param cmdResult
     */
    public static void PlayerPrint(Player player, CmdResult<String> cmdResult) {
        // result name 
        String result = cmdResult.toString();
        if(cmdResult.type ==  203) {
            //GetPluginApp().getDescription().getName()
            result = result.replaceAll("%name%", PluginEntity.getPluginDescrption().getName());
        }
        if(cmdResult.type == 102) {
            result = RanbowString(result);
        }
 
        player.sendMessage(ChatColor.translateAlternateColorCodes('&', result));
    }

    /**
     * send  Message to mulit-player
     * @param players
     * @param cmdResult
     */
    public static void PlayerPrint(Collection<? extends Player> players, CmdResult<String> cmdResult) {
        // result name 
        for(Player ply : players) {
            PlayerPrint(ply,cmdResult);
        }
    }

    public static String RanbowString(String rawString){


        StringBuffer buffer = new StringBuffer();
        Random r = new Random();
        //boolean check_vailable = false;

        for(char ch :rawString.toCharArray()){
            String randomColour = ConfigKV.CHAT_COLOR[r.nextInt(21)]; //You need to do the get random colour part ;)
            //bai(int)Math.random() * ConfigKV.CHAT_COLOR.length 
            //chat_color total type = 21

            buffer.append(randomColour + ch);
        }
        return buffer.toString();
    }




	public T getData() {
		return data;
	}
	public void setData(T data) {
		this.data = data;
	}
	public int getType() {
		return type;
	}
	public void setType(int type) {
		this.type = type;
	}
	public String getTitle() {
		return title;
	}
	public void setTitle(String title) {
		this.title = title;
	}

    @Override
    public String toString() {
        String result = "";
        if(title.length()>3){
            result = title+" : ";
            
        }
        result = result + data.toString();
        return result;
    }
}

File: src/main/kotlin/cc/squirtle/playerstatistics/entity/CmdResult.kt
package cc.squirtle.playerstatistics.entity

import cc.squirtle.util.DealString
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*
import java.lang.*

class CmdResult<T>
    (type: Int?, title: String?, data: T?){
    private var type: Int? = type
    private var title: String? = title
    private var data: T? = data

    override fun toString():String {
        var msg :String
        when(this.type){
            203 -> {
                this.title = this.title!!.replace("%plugin_name%".toRegex(), PluginEntity.PLUGIN_DESCFILE!!.name)
                msg = "${this.title} ${this.data}"
            }
            102 -> {
                msg = "${this.title} ${this.data}"
                msg = DealString.RanbowString(msg)
            }
            else -> {
                msg = "${this.title} ${this.data}"
            }
        }

        return msg
    }

    /**
     * send message to console
     * ex:Entity.SUCCESS("success").Send2Console()
     */
    fun Send2Console(){
        val result: String? = this.toString()
        Bukkit.getConsoleSender().sendMessage(ChatColor.translateAlternateColorCodes('&', result))
    }




    /**
     * send Message to single-player
     * @param player
     */
    fun Send2Player(player: Player) {
        // result name
        val result = this.toString()
        player.sendMessage(ChatColor.translateAlternateColorCodes('&', result))
    }

    /**
     * send  Message to mulit-player
     * @param players
     */
    fun Send2Player(players: Collection<Player>) {
        for (ply in players) {
            Send2Player(ply)
        }
    }




companion object {

    fun <T> SUCCESS(data: T): CmdResult<T> {
        return CmdResult(ResType.SUCCESS.type,ResType.SUCCESS.title,data)
    }

    fun <T> FAILED(data: T): CmdResult<T> {
        return CmdResult(ResType.FAILED.type,ResType.FAILED.title,data)
    }

    fun <T> NOTICE(data: T): CmdResult<T> {
        return CmdResult(ResType.NOTICE.type, ResType.NOTICE.title, data)
    }

    fun <T> INFO(data: T): CmdResult<T> {
        return CmdResult(ResType.INFO.type, ResType.INFO.title, data)
    }

    fun <T> NAME(data: T): CmdResult<T> {
        return CmdResult(ResType.NAME.type, ResType.NAME.title, data)
    }

    fun <T> UNAVAIL(data: T): CmdResult<T> {
        return CmdResult(ResType.UNAVAIL.type, ResType.UNAVAIL.title, data)
    }

    fun <T> RAINBOW(data: T?): CmdResult<T> {
        return CmdResult(ResType.RAINBOW.type, ResType.RAINBOW.title, data)
    }

    /**
     * send Message to single-player
     * @param player
     */
    fun <T> Send2Player(player: Player, msg: CmdResult<T>) {
        // result name
        val result = msg.toString()
        player.sendMessage(ChatColor.translateAlternateColorCodes('&', result))
    }

    /**
     * send  Message to mulit-player
     * @param players
     * @param cmdResult
     */
    fun <T> Send2Player(players: Collection<Player>, cmdResult: CmdResult<T>) {
        for (ply in players) {
            Send2Player(ply, cmdResult)
        }
    }



    /**
     * send message to console
     * ex:Entity.Send2Console(Entity.SUCCESS("success"))
     */
    fun <T> Send2Console(msg: CmdResult<T>){
        val result = msg.toString()
        Bukkit.getConsoleSender().sendMessage(ChatColor.translateAlternateColorCodes('&', result))
    }


}



}
